### ISSUE TITLE
RFC: design for efficient lookup of old names

### ISSUE THREAD

**Issue by pchiusano** [unisonweb/unison#1346]
In the repo format, add the following directory:

```
names/<hash1-2>/<hash3-4>/.../<hash511-512/
  namehash1.name
  namehash2.name
  ...
```

Where `<hash1-2>` are the first two chars of the hash, `<hash3-4>` are the 3rd and 4th characters of the hash, etc. `namehash1.name` is a hash of the name and the file contents are a UTF-8 encoded `Name`.

So it's an append-only, on-disk `Trie HashChar (Set Name)`. Whenever definitions are added to the codebase, they are inserted into this trie. Supports efficient lookup of any partial hash. Merges via file system operations efficiently without copying any data into Haskell memory.

The way this would be used:

* Before pretty-printing anything, any references not covered by current namespace would get looked up in the names index. These would get used as the second`Names3` and get hash qualified via the existing mechanism. So this replaces the historical search through the branch that is sometimes super slow.
* Suggest that these get synced via `push` / `pull`, with the behavior that on `pull`, if a name already exists for a reference, names from the source are ignored, and if pushing, only fill in a name if none exists for a hash.
* Later, we could potentially expose UCM commands for maintenance of this names index, like updating the indexed name / picking a winner.

**Comment by aryairani** [comment]
How do we add this to codebases that are currently missing it?  

Option 1: Build from scratch if it's missing, and then once it exists, assume it is complete?  
Option 2: Include a list of namespace hashes that are represented in the index.  Start indexing from the head namespace.  To index a namespace `ns`:  If `hash ns` is already in the index, then you're done.  Otherwise, push `ns` onto a stack, index `predecessor ns`, add the names from `ns` (optionally delete older names for the same hash), add `hash ns`, pop the stack and continue.

**Comment by aryairani** [comment]
Superseded by #2128
